src/theory/quantifiers/ceg_t_instantiator.h

```diff
@@ -262,7 +265,41 @@ class BvInstantiator : public Instantiator {
                       Node pv,
                       Node lit,
                       Node alit,
-                      InstEffort effort);
+                      CegInstEffort effort);
+};
+
+/** Bitvector instantiator preprocess
+ *
+ * This class implements preprocess techniques that are helpful for
+ * counterexample-guided instantiation, such as introducing variables
+ * that refer to disjoint bit-vector extracts.
+ */
+class BvInstantiatorPreprocess : public InstantiatorPreprocess
+{
+ public:
+  BvInstantiatorPreprocess() {}
+  virtual ~BvInstantiatorPreprocess() {}
+  /** register counterexample lemma
+   *
+   * This method modifies the contents of lems based on removing extract terms
+   * when the option --cbqi-bv-rm-extract is enabled, for example:
+   *   P[ ((extract 7 4) t), ((extract 3 0) t)]
+   *     becomes:
+   *   P[x,y] ^ t = concat( x, y )
+   * where x and y are fresh variables.
+   */
+  virtual void registerCounterexampleLemma(std::vector<Node>& lems,
+                                           std::vector<Node>& ce_vars) override;
+
+ private:
+  /** map from terms to bitvector extracts applied to that term */
+  std::map<Node, std::vector<Node> > d_extract_map;
+  /** process
+   * This method collects all extract terms in lem
+   * and stores them in d_extract_map.
+   * visited is the terms we've already visited.
+   */
+  void process(Node lem, std::unordered_set<TNode, TNodeHashFunction>& visited);
```

Comment:
`process` is a bit misleading here. Since this is only used for finding extracts in lem, maybe call it `collect_extracts`?